web3/manager.py

```diff
@@ -198,43 +208,66 @@ async def _coro_make_request(
         self.logger.debug(f"Making request. Method: {method}")
         return await request_func(method, params)
 
+    #
+    # formatted_response parses and validates JSON-RPC responses
+    # for expected properties (result or an error) with the expected types.
+    # Responses are inconsistent so allow for missing required JSON-RPC properties
+    #
     @staticmethod
     def formatted_response(
         response: RPCResponse,
         params: Any,
         error_formatters: Optional[Callable[..., Any]] = None,
         null_result_formatters: Optional[Callable[..., Any]] = None,
     ) -> Any:
+        if "jsonrpc" in response and response["jsonrpc"] != "2.0":
+            _raise_bad_response_format(response, '"jsonrpc" must equal "2.0"')
+
+        if "id" in response and not isinstance(response["id"], (str, int, None)):
```

Comment:
This raises a warning for me coming from the `type` for `id` in `RPCResponse`. We define it as an `int` only. I think the `id` can be a string so maybe we need to change the type there to `Union[int, str]` to match. 

I also don't think we should be validating a value of `None` here so loosely. Per the JSON-RPC 2.0 spec:

> id
>     This member is REQUIRED.
>     It MUST be the same as the value of the id member in the Request Object.
>     **If there was an error in detecting the id in the Request object (e.g. Parse error/Invalid Request), it MUST be Null.**

I think we grab the `id` and if it's `None` we make sure that it there was an `error` as well.

```py
        if "id" in response:
            response_id = response.get("id")
            if response_id is None and "error" not in response:
                _raise_bad_response_format(
                    response,
                    '"id" was null without error in response.',
                )
            elif not isinstance(response_id, (str, int)):
                _raise_bad_response_format(
                    response,
                    '"id" must be a string or integer.',
                )
```

This won't actually fail any of the current tests, so we should add a test that includes an `"id"` key in the response but does not include an `"error"` and validate the new exception is raised.

Unfortunately subscriptions in ethereum don't include an `id` in the response so this actually deviates from JSON-RPC 2.0. Which is why we keep the `if "id" in response` rather than making sure that it is always there. And if it's in the response it can only be null when there is an error.

Thoughts?

Reply:
Responses with an id may contain an error or a result so no need to enforce the check for an error within the "id" block.

I also saw a [note in the "id" section of the response object](https://www.jsonrpc.org/specification#response_object) of the spec that the value must be null when the request throws a parse error or invalid request. Unfortunately a bunch of the response types appear to lack an "id" entirely and seems like we are stuck with it deviating from the spec in this case.

The specific case you called out is caught in the final `else` of `format_response`, which is covered by a few tests like https://github.com/ethereum/web3.py/blob/main/tests/core/manager/test_response_formatters.py#L31

Reply:
My brain wasnt groking the error rule for some reason. It sounds good to me to enforce that case.

I read it as "if there is an error, the value of id must be NULL", therefore if id is anything but "null" it should raise an exception? Curious if we want to keep empty string and None as empty values here, especially since empty string could mean something different.

The spec for the requests also indicates if id is missing, the request is considered to be a message. I'm guessing the same is for responses and the way the logic is implemented it will be parsed the same as other errors, so no change needed for that case.

Reply:
> I read it as "if there is an error, the value of id must be NULL", therefore if id is anything but "null" it should raise an exception?

Yeah I think we're on the same page. If `id` key is present and `error` is present, `id` value must be `null`. My only question is what do they mean by `null`. They use `NULL` and `Null` in the text and `null` (not the string `"null"`) in the examples below. My understanding is that json parsing into a Python object would yield `None` in those cases. So that's probably the only reasonable `null` value? Thoughts there?

Since the `id` can indeed be `None` when there's an `error` also present, that makes the typing an `Optional[...]`.

If we're validating that, it could just look like the example block I posted above where we check if it is indeed `None` and there is no `error`, we throw... and then we only check for `str` or `int` cases below that since it can't be `None` beyond that first bit of logic. Does that make sense?

Reply:
Some new test cases arise from this.

1. There is an `"id"` key in the response with a value of `None` and no `"error"` key -> raises exception
2. There is an `"id"` key in the response with a value other than `None` while an `"error"` key is also present -> raises exception
3. There is an `"id"` key in the response witha  value of `None` and there is an `error` key with properly formatted `error` object -> does NOT raise exception